### PR TITLE
Feature/mai finance

### PR DESCRIPTION
**Pull request by majkic99**
npx hardhat changeRepoNetwork Mainnet Optimism
npx hardhat node (with optimism rpc)
npx hardhat test --network localOptimism test/qidao/...

https://github.com/defisaver/defisaver-sdk/pull/45

### CODE REVIEW

File: contracts/actions/qidao/QiDaoGenerate.sol [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    /// @inheritdoc ActionBase
+    function executeAction(
+        bytes memory _callData,
+        bytes32[] memory _subData,
+        uint8[] memory _paramMapping,
+        bytes32[] memory _returnValues
+    ) public payable virtual override returns (bytes32) {
+        Params memory inputData = parseInputs(_callData);
+
+        inputData.userVaultId = _parseParamUint(
```
Comment (kjaredic) [inline contracts/actions/qidao/QiDaoGenerate.sol:30]: why not map vaultId as well? could be useful for strategies idk

File: contracts/views/QiDaoView.sol [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.10;
+
+import "../actions/qidao/helpers/QiDaoHelper.sol";
+
+contract QiDaoView is QiDaoHelper {
+}
```
Comment (kjaredic) [inline contracts/views/QiDaoView.sol:8]: It's a view but i see nothing

File: contracts/actions/qidao/QiDaoSupply.sol [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        internal
+        returns (uint256, bytes memory logData)
+    {
+        if (_inputParams.from == address(0)) revert NullAddressTransfer();
+
+        address vaultAddress = vaultRegistry.vaultAddressById(_inputParams.vaultId);
+        address collateralAsset = IStablecoin(vaultAddress).collateral();
+
+        collateralAsset.pullTokensIfNeeded(_inputParams.from, _inputParams.amount);
+        collateralAsset.approveToken(vaultAddress, _inputParams.amount);
```
Comment (kjaredic) [inline contracts/actions/qidao/QiDaoSupply.sol:79]: if amount is uintMax this will give max approval

**Review by kjaredic — COMMENTED** [review]
looks good, see about mapping vaultId

**Review by kjaredic — APPROVED** [review]
lgtm

File: contracts/actions/qidao/QiDaoPayback.sol [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        return uint8(ActionType.STANDARD_ACTION);
+    }
+
+    //////////////////////////// ACTION LOGIC ////////////////////////////
+
+    function _qiDaoPayback(Params memory _inputParams)
+        internal
+        returns (uint256, bytes memory logData)
+    {
+        if (_inputParams.from == address(0)) revert NullAddressTransfer();
```
Comment (yippee-ki-yay) [inline contracts/actions/qidao/QiDaoPayback.sol:75]: No need to check _from == 0x0 as we can't pull tokens from 0x0

File: contracts/actions/qidao/QiDaoPayback.sol [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+        address vaultAddress = vaultRegistry.vaultAddressById(_inputParams.vaultId);
+
+        if (_inputParams.amount == type(uint256).max) {
```
Comment (yippee-ki-yay) [inline contracts/actions/qidao/QiDaoPayback.sol:79]: Take userVault whole debt when the amount sent is bigger than the userVault debt to have the same behaviour as Maker

File: contracts/actions/qidao/QiDaoSupply.sol [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        return uint8(ActionType.STANDARD_ACTION);
+    }
+
+    //////////////////////////// ACTION LOGIC ////////////////////////////
+
+    function _qiDaoSupply(Params memory _inputParams)
+        internal
+        returns (uint256, bytes memory logData)
+    {
+        if (_inputParams.from == address(0)) revert NullAddressTransfer();
```
Comment (yippee-ki-yay) [inline contracts/actions/qidao/QiDaoSupply.sol:76]: Same no need to check _from == 0x0

File: contracts/actions/qidao/QiDaoSupply.sol [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        address vaultAddress = vaultRegistry.vaultAddressById(_inputParams.vaultId);
+        address collateralAsset = IStablecoin(vaultAddress).collateral();
+
+        _inputParams.amount = collateralAsset.pullTokensIfNeeded(_inputParams.from, _inputParams.amount);
```
Comment (yippee-ki-yay) [inline contracts/actions/qidao/QiDaoSupply.sol:81]: Handle uint.max take the whole balance of _from

File: contracts/actions/qidao/helpers/QiDaoRegistry.sol [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    mapping(uint16 => address) public vaultAddressById;
+
+    uint16 public latestVaultId = 0;
+
+    function addVault(address _vaultAddr) public onlyOwner returns (uint16 vaultId) {
+        if (vaultIdByVaultAddress[_vaultAddr] != 0) revert VaultAlreadyRegistered(vaultIdByVaultAddress[_vaultAddr]);
+        vaultId = registerVault(_vaultAddr);
+    }
+
+    function registerVault(address _vaultAddr) internal returns (uint16 vaultId){
```
Comment (yippee-ki-yay) [inline contracts/actions/qidao/helpers/QiDaoRegistry.sol:20]: Just a preference but putin the internal method below the removeVault which has permission of onlyOwner

File: contracts/actions/qidao/helpers/QiDaoRegistry.sol [hunk 9]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.10;
+
+import "../../../auth/AdminAuth.sol";
+
+contract QiDaoRegistry is AdminAuth {
```
Comment (yippee-ki-yay) [inline contracts/actions/qidao/helpers/QiDaoRegistry.sol:6]: Comments on methods and what the contract is for

File: contracts/actions/qidao/helpers/QiDaoRegistry.sol [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    error VaultAlreadyRegistered(uint16);
+
+    /// @notice IDs start from 1
+    mapping(address => uint16) public vaultIdByVaultAddress;
+    mapping(uint16 => address) public vaultAddressById;
```
Comment (yippee-ki-yay) [inline contracts/actions/qidao/helpers/QiDaoRegistry.sol:11]: Maybe a check when we take vaultAddr as they could be 0x0, currently it isn't a problem as the code will just fail, but a low lvl. call to 0x0 might pass as success and leave some tokens on proxy etc...

File: test/actions.js [hunk 11]
```diff
@@ -32,14 +32,14 @@ const { getSecondTokenAmount } = require('./utils-uni');
 const { LiquityActionIds, getHints, getRedemptionHints } = require('./utils-liquity');
 const { execShellCommand } = require('../scripts/hardhat-tasks-functions');
 
-const executeAction = async (actionName, functionData, proxy, regAddr = REGISTRY_ADDR) => {
```
Comment (yippee-ki-yay) [inline test/actions.js:35]: removing regAddr = REGISTRY_ADDR breaks forkoor related stuff which need to inject registry like mcdWithdraw action

File: test/qidao/qidao-tests.js [hunk 12]
```diff
@@ -0,0 +1,535 @@
+const hre = require('hardhat');
```
Comment (yippee-ki-yay) [inline test/qidao/qidao-tests.js:1]: Test code paths related to uint-max flow

File: contracts/views/QiDaoView.sol [hunk 13]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.10;
+
+import "../actions/qidao/helpers/QiDaoHelper.sol";
+import "../interfaces/qidao/IStablecoin.sol";
+
+contract QiDaoView is QiDaoHelper {
+
+    function getVaultInfo(uint16 _vaultId, uint256 _userVaultId) public view returns (uint256 coll, uint256 debt) {
```
Comment (yippee-ki-yay) [inline contracts/views/QiDaoView.sol:10]: Not just here but maybe in general don't use `_vaultId` as we are use to meaning a diff. thing in Maker. Instead something like qiVaultId, qiDaoVaultId, collVaultId, etc... something to indicate that it's a protocol wide id.

**Review by yippee-ki-yay — CHANGES_REQUESTED** [review]
In general the implementation looks good, can further optimize calldata savings and there are a few more edge cases we can cover (notes in comments).

**Pull request merged**